pkg/util/pdag/pdag.go

```diff
@@ -0,0 +1,194 @@
+// Copyright 2025, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package pdag provides facilities for constructing a DAG and the traversing it
+// efficiently in parallel.
+package pdag
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+
+	"golang.org/x/sync/errgroup"
+)
+
+type DAG[T any] struct {
+	nodes   []node[T]
+	edges   []edge
+	adjList [][]int // adjacency list: adjList[i] contains indices of nodes reachable from node i
```

Comment:
I think this could use some clarification. if I'm reading the code correctly, each entry isn't a complete reachability list--it's more entry `i` is a list of nodes to which node `i` has an edge (so node `i`'s outgoing neighbors). is that right?

related--I'm curious why you chose the global edge list rather than edges-as-pointers or a per-node edge list. feels like a per-node edge list would suffice.

Reply:
It would. It mostly just evolved that way here. I can change it to per-node. I think that's a bit cleaner.